### ISSUE TITLE
Why are OrderedDicts used in the model interface?

### ISSUE THREAD

**Issue by matthiaskoenig** [biosustain/optlang#98]
Hi all,
I have to run a lot of FBA simulations and looking for places to speed them up. I am mainly interested in getting the reduced_costs, constraint_values, shadow_prices, ... as fast as possible from the model.

I had look at the interface.Model and wondered myself why are there OrderedDicts used for this values (in comparison to simple dict). This creates unnecessary overhead, costs performance and I don't see any benefits to it. 
Or are there any algorithms relying on the order of the variables or constraint ids ?
If not all the OrderedDicts should be replaced with simple dicts in the interface.

In addition you should hash the names of your variables, constraints, ...
On every query of reduced costs, ... you reiterate and recalculate for instance the list of variable names
(variable.name for variable in self.variables)
There should be some property/field 
`variable_names`, `constraint_names`
on the Model class which are just updated when new variables are added/removed.

```
    @property
    def reduced_costs(self):
        if self.is_integer:
            raise ValueError("Dual values are not well-defined for integer problems")
        try:
            return collections.OrderedDict(
                zip((variable.name for variable in self.variables), self.problem.solution.get_reduced_costs()))
        except CplexSolverError as err:
            raise SolverError(str(err))

    @property
    def constraint_values(self):
        try:
            return collections.OrderedDict(
                zip((constraint.name for constraint in self.constraints), self.problem.solution.get_activity_levels()))
        except CplexSolverError as err:
            raise SolverError(str(err))
```
M

**Comment by matthiaskoenig** [comment]
https://stackoverflow.com/questions/25056387/ordereddict-vs-dict-in-python
~ speedup of 7-8x for creation dict vs. orderdDict
But the opinions/benchmarks differ. But if not needed for anything other than we have the ids ordered, than drop the OrderedDicts (the very few use cases where one needs ordered keys people can sort).

**Comment by matthiaskoenig** [comment]
Also it looks like there are functions to get all the names at once
`self.problem.variables.get_names()`
instead of 
`(variable.name for variable in self.variables)`

**Comment by KristianJensen** [comment]
Regarding the ordered dicts, I think the only reason for using them is for convenience, and changing them to dict would not backwards compatible. Have you done any benchmarks that show that construction of the OrderedDict is the bottleneck, as opposed to actually getting the values from the solver?

Regarding caching the names I doubt that the attribute lookup performed when getting the name of a var/constraint makes a practical difference. A quick test shows me that each attribute lookup takes on the order of 10 ns. This is not worth trading for the extra complexity of maintaining a list of names, imo.

**Comment by matthiaskoenig** [comment]
I did some benchmarks for some similar thing and saw a factor of 10 for OrderedDicts compared to Dicts (see below). 
Backwards compatibility I don't understand. OrderedDicts behave like dicts in almost every context (iteration, access).
Only difference is that the keys are ordered, but I did not see any code which relies on that the keys of the dict are ordered (at least in cobrapy).

Here some quick benchmark.
```
import cobra
from collections import OrderedDict
from pandas import Series
sbml_path = "iAB_RBC_283.xml"
model = cobra.io.read_sbml_model(sbml_path)
model.optimize()

var_primals = model.solver.primal_values
rxn_index = [rxn.id for rxn in model.reactions]
data = [var_primals[rxn.id]-var_primals[rxn.reverse_id] for rxn in model.reactions]
```

```
%timeit -n 1000 fluxes = dict(zip(rxn_index, data))
1000 loops, best of 3: 48.3 µs per loop

%timeit -n 1000 fluxes = OrderedDict(zip(rxn_index, data))
1000 loops, best of 3: 519 µs per loop
```

**Comment by KristianJensen** [comment]
You are only benchmarking the construction of the dicts themselves. How does this compare to the total time it takes to get the primals from the solver?

**Comment by matthiaskoenig** [comment]
Hi @KristianJensen, 
I looked a bit at the code and would like to contribute some functions to the Model interface.

My real problem is mainly that I can't get via optlang the arrays with 
`var_primals` and `var_duals`
in a fast way (the OrderedDict is just a symptom of the underlying main problem).

I would like to contribute on interface.Model (and cplex_interface.Model / glpk_interface.Model) the functions:
```
def primal_values_array(self):
def reduced_costs_array(self):
def variables_ids(self):
```
which directly return the array of the respective solvers (making use of the solver vector functions to return these arrays)
and a function which allows to get the list of variables from the model. Than for a given problem I can just query the variable ids once and get very fast the solution arrays.
Than I wouldn't have to interact with OrderedDicts and avoid the element wise copy of numerical values from the arrays which are performed currently.

Is this okay?

**Comment by KristianJensen** [comment]
I was actually thinking of refactoring the value retrievals in a similar way. If you want to give it a go, that would be very welcomed.

The refactoring would be the following (very close to what you describe):
 - In each solver interface, the Model class would have the methods `_get_primal_values()`, `_get_constraint_values()`, `_get_reduced_costs()` and `_get_shadow_prices()`
- These methods would just directly return whatever array-like object the solver returns.
- The current properties (`primal_values`, `shadow_prices` etc) could then be moved to `interface.py` and call the respective `_get_{something}` methods.

The idea behind this is as you describe, that someone who is very concerned with performance can use these methods directly. 

I don't see a reason to add a `variable_ids` methods since, as I mentioned earlier, just doing `[var.name for var in model.variables]` will be very fast.

**Comment by Midnighter** [comment]
Just a small point, if you do intend for the `get` methods to be used (albeit by a small number of users), I would vote for dropping the initial underscore marking them as private as they are intended to be part of the public class interface.

**Comment by KristianJensen** [comment]
@Midnighter For now I would like to keep the underscore, exactly so that they're not part of the public interface.

**Comment by matthiaskoenig** [comment]
I will try to implement this. 
How should I call the respective id functions
`_get_variables_ids()`, `get_constraints_ids()` ?

**Comment by matthiaskoenig** [comment]
with underscore both: `_get_variables_ids()`, `_get_constraints_ids()` ?

**Comment by matthiaskoenig** [comment]
I have a first version in
https://github.com/biosustain/optlang/pull/101
still have to fix up some smaller things, but this is how I would change it.
Feedback welcome.

This allows easily to get the vectorized primals, duals, ... for glpk and cplex. 
I would like to use numpy arrays but did not want to add the dependency, so going for lists right now.
Let me know if it is okay to add numpy.

**Comment by matthiaskoenig** [comment]
This can be closed with merge of #101.
Everything answered.
